src/Plutarch/Extra/IsData.hs

```diff
@@ -0,0 +1,168 @@
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE QuantifiedConstraints #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Plutarch.Extra.IsData (
+    -- * PlutusTx ToData/FromData derive-wrappers
+    ProductIsData (..),
+    EnumIsData (..),
+
+    -- * Plutarch PIsData/PlutusType derive-wrappers
+    PEnumData (..),
+    PConstantViaDataList (PConstantViaDataList),
+
+    -- * Functions for PEnumData types
+    pmatchEnum,
+    pmatchEnumFromData,
+) where
+
+--------------------------------------------------------------------------------
+
+import Data.Coerce (coerce)
+import Data.Proxy (Proxy (..))
+import Generics.SOP (
+    All,
+    IsProductType,
+    hcmap,
+    hcollapse,
+    hctraverse,
+    mapIK,
+    mapKI,
+    productTypeFrom,
+    productTypeTo,
+    unI,
+ )
+import qualified Generics.SOP as SOP
+import Plutarch (PlutusType (pcon', pmatch'))
+import Plutarch.Builtin (PIsData (pdataImpl, pfromDataImpl), pasInt)
+import Plutarch.Extra.TermCont (pletC)
+import Plutarch.Lift (PConstantDecl (..))
+import Plutarch.Prelude
+import Plutarch.Unsafe (punsafeCoerce)
+import PlutusLedgerApi.V1 (BuiltinData (BuiltinData))
+import PlutusTx (Data (List), FromData (..), ToData (..), fromData, toData)
+import Prelude
+
+--------------------------------------------------------------------------------
+-- ProductIsData
+
+{- |
+  Wrapper for deriving 'ToData', 'FromData' using the 'List' constructor of Data to represent a Product type.
+
+  Uses 'gProductToBuiltinData', 'gproductFromBuiltinData'.
+-}
+newtype ProductIsData a = ProductIsData {unProductIsData :: a}
+
+-- | Variant of 'PConstantViaData' using the List repr from 'ProductIsData'
+newtype PConstantViaDataList (h :: Type) (p :: PType) = PConstantViaDataList h
+
+-- | Generically convert a Product-Type to 'BuiltinData' with the 'List' repr
```

Comment:
Nitpick, but. When it comes to haddock docstrings, we've been ending them with a period everywhere.